src/algorithm/const_fold.rs

```diff
@@ -0,0 +1,302 @@
+//! Constant folding routines.
+
+use std::collections::{BTreeSet, HashMap};
+
+use itertools::Itertools;
+
+use crate::{
+    builder::{DFGBuilder, Dataflow, DataflowHugr},
+    extension::{ConstFoldResult, ExtensionRegistry},
+    hugr::{
+        rewrite::consts::{RemoveConst, RemoveConstIgnore},
+        views::SiblingSubgraph,
+        HugrMut,
+    },
+    ops::{Const, LeafOp, OpType},
+    type_row,
+    types::{FunctionType, Type, TypeEnum},
+    values::Value,
+    Hugr, HugrView, IncomingPort, Node, SimpleReplacement,
+};
+
+/// Tag some output constants with [`OutgoingPort`] inferred from the ordering.
+fn out_row(consts: impl IntoIterator<Item = Const>) -> ConstFoldResult {
+    let vec = consts
+        .into_iter()
+        .enumerate()
+        .map(|(i, c)| (i.into(), c))
+        .collect();
+    Some(vec)
+}
+
+/// Sort folding inputs with [`IncomingPort`] as key
+fn sort_by_in_port(consts: &[(IncomingPort, Const)]) -> Vec<&(IncomingPort, Const)> {
+    let mut v: Vec<_> = consts.iter().collect();
+    v.sort_by_key(|(i, _)| i);
+    v
+}
+
+/// Sort some input constants by port and just return the constants.
+pub(crate) fn sorted_consts(consts: &[(IncomingPort, Const)]) -> Vec<&Const> {
+    sort_by_in_port(consts)
+        .into_iter()
+        .map(|(_, c)| c)
+        .collect()
+}
+/// For a given op and consts, attempt to evaluate the op.
+pub fn fold_const(op: &OpType, consts: &[(IncomingPort, Const)]) -> ConstFoldResult {
+    let op = op.as_leaf_op()?;
+
+    match op {
+        LeafOp::Noop { .. } => out_row([consts.first()?.1.clone()]),
+        LeafOp::MakeTuple { .. } => {
+            out_row([Const::new_tuple(sorted_consts(consts).into_iter().cloned())])
+        }
+        LeafOp::UnpackTuple { .. } => {
+            let c = &consts.first()?.1;
+
+            if let Value::Tuple { vs } = c.value() {
+                if let TypeEnum::Tuple(tys) = c.const_type().as_type_enum() {
+                    return out_row(tys.iter().zip(vs.iter()).map(|(t, v)| {
+                        Const::new(v.clone(), t.clone())
+                            .expect("types should already have been checked")
+                    }));
+                }
+            }
+            None // could panic
```

Comment:
I am not sure the comment is correct, i.e. what would `panic` and how returning None would help. I suppose you mean that the `let if` clauses might fail?

Reply:
i mean that `None` shouldn't ever be returned here, so it might be clearer to panic